learn/building_programs.md

```diff
@@ -0,0 +1,780 @@
+
+Managing libraries (static and dynamic libraries)
+-------------------------------------------------
+
+If you need to manage a program built of dozens of source files (and
+that is not uncommon!), the command-line needed to specify all the
+object files will be very long indeed. This soon becomes tedious or
+even impossible to maintain. So a different solution is called for:
+create your own libraries.
+
+Libraries contain any number of object files in a compact form, so that
+the command-line becomes far shorter:
+
+    $ gfortran -o tabulate tabulate.f90 function.o supportlib.a
+
+where "supportlib.a" is a collection of one, two or many object files,
+all compiled and then put into a library. The extension ".a" is used by
+Linux and Linux-like platforms. On Windows the extension ".lib" is used.
+
+Creating your own libraries is not that complicated: you use a utility
+like "ar" or (on Windows) "lib" to achieve this:
+
+    $ gfortran -c file1.f90 file2.f90
+    $ gfortran -c file3.f90 ...
+    $ ar r supportlib.a file1.o file2.o
+    $ ar r supportlib.a file3.o ...
+
+or, using the "link" utility:
+
+    c:\...> ifort -c file1.f90 file2.f90
+    c:\...> ifort -c file3.f90 ...
+    c:\...> lib /out:supportlib.lib file1.obj file2.obj
+    c:\...> link supportlib.lib file3.obj ...
+
+Note:
+
+* The command "ar" with the option "r" either creates the library (the
+name appears after the option) or adds new object files to the library
+(or replaces any existing ones).
+* The command "lib" will create a new library if you use specify the
+option "/out:" with the name of the new library next to it. To add
+object files to an existing library, leave out the "/out:" bit.
+* On platforms like Linux there is a particular convention to name
+libraries. If you name your library like "libname.a" (note the "lib"
+prefix), then you can refer to it as "-lname" in the link step.
+* Libraries are often sought in directories indicated by an option "-L"
+or "/LIBPATH". This saves you from having to specify the exact path for
+every library.
+
+Using libraries you can build very large programs without
+having to resort to extremely long command lines.
+
+
+### Static versus dynamic libraries
+
+The above discussion is tacitly assuming that you are using so-called
+static libraries. Static libraries (or at least parts of their
+contents) become an integral part of the executable program. The only
+way to change the routines incorporated in the program is by rebuilding
+the program with a new version of the library.
+
+A flexible alternative is to use so-called dynamic libraries. These
+libraries remain outside the executable program and as a consequence you
+can replace them without rebulding the entire program. Compilers and
+indeed the operating system itself rely heavily on such dynamic
+libraries. You could consider dynamic libraries as a sort of executable
+programs that need a bit of help to be run.
+
+Building dynamic libraries works slightly differently:

```

Comment:
Differently based on what? I have the feeling that something is missing in this sentence.